File: contract-service/src/domain/ContractDraft.ts
import { v4 as uuidv4 } from 'uuid';

/**
 * Entity: ContractDraft
 * ร่างสัญญาจะซื้อจะขาย — สร้างเมื่อ KYC ผ่านและ booking confirmed
 * อยู่ภายใน Contract Aggregate access ผ่าน Contract Root เท่านั้น
 */
export class ContractDraft {
  draftId: string;
  fileUrl: string;
  templateId: string;
  draftedAt: string;

  private constructor(draftId: string, templateId: string, fileUrl: string, draftedAt: string) {
    this.draftId = draftId;
    this.templateId = templateId;
    this.fileUrl = fileUrl;
    this.draftedAt = draftedAt;
  }

  static create(templateId: string, fileUrl: string): ContractDraft {
    return new ContractDraft(uuidv4(), templateId, fileUrl, new Date().toISOString());
  }

  static rehydrate(draftId: string, templateId: string, fileUrl: string, draftedAt: string): ContractDraft {
    return new ContractDraft(draftId, templateId, fileUrl, draftedAt);
  }
}

File: _java-archive/kafka-listener/src/main/java/dev/kritchalach/kafka/domain/ContractDraft.java
package dev.kritchalach.kafka.domain;

import jakarta.persistence.*;
import java.time.Instant;
import java.util.UUID;

/**
 * Entity: ContractDraft
 * เก็บข้อมูลของร่างสัญญาจะซื้อจะขาย สร้างขึ้นเมื่อ KYC ผ่านและ booking confirmed
 * อยู่ภายใน Contract Aggregate — ไม่ถูก access โดยตรงจากภายนอก ต้องผ่าน Contract Root
 */
@Entity
@Table(name = "contract_drafts")
public class ContractDraft {

    @Id
    @GeneratedValue(strategy = GenerationType.UUID)
    private UUID draftId;

    private String fileUrl;

    private UUID templateId;

    @Column(nullable = false)
    private Instant draftedAt;

    protected ContractDraft() {}

    public ContractDraft(UUID templateId, String fileUrl) {
        this.templateId = templateId;
        this.fileUrl = fileUrl;
        this.draftedAt = Instant.now();
    }

    public UUID getDraftId() { return draftId; }
    public String getFileUrl() { return fileUrl; }
    public UUID getTemplateId() { return templateId; }
    public Instant getDraftedAt() { return draftedAt; }
}
